pkg/ring/lifecycler.go

```diff
@@ -343,14 +345,52 @@ func (i *Lifecycler) setTokens(tokens Tokens) {
 	i.stateMtx.Lock()
 	defer i.stateMtx.Unlock()
 
-	i.tokens = tokens
+	i.tokenFile.Tokens = tokens
 	if i.cfg.TokensFilePath != "" {
-		if err := i.tokens.StoreToFile(i.cfg.TokensFilePath); err != nil {
+		if err := i.tokenFile.StoreToFile(i.cfg.TokensFilePath); err != nil {
 			level.Error(i.logger).Log("msg", "error storing tokens to disk", "path", i.cfg.TokensFilePath, "err", err)
 		}
 	}
 }
 
+func (i *Lifecycler) getPreviousState() InstanceState {
+	i.stateMtx.RLock()
+	defer i.stateMtx.RUnlock()
+	return i.tokenFile.PreviousState
+}
+
+func (i *Lifecycler) setPreviousState(state InstanceState) {
+	i.stateMtx.Lock()
+	defer i.stateMtx.Unlock()
+
+	if !(state == ACTIVE || state == READONLY) {
+		level.Error(i.logger).Log("msg", "cannot store unsupported state to disk", "new_state", state, "old_state", i.tokenFile.PreviousState)
```

Comment:
Is it possible to log this for other states in real scenarios? I don't see we filter out other states at the caller side so wondering if we will produce some unnecessary error level logs.

Reply:
When stopping, it would change state to `LEAVING` (https://github.com/cortexproject/cortex/blob/master/pkg/ring/lifecycler.go#L568). Based on logic here: https://github.com/cortexproject/cortex/blob/master/pkg/ring/lifecycler.go#L911, only `ACTIVE` or `READONLY` state can be changed to `LEAVING`. So, in theory, this error log should not be called if `setPreviousState` is only called when stopping.